libraries/AC_AttitudeControl/AC_AttitudeControl_Multi.h

```diff
@@ -60,6 +60,9 @@ class AC_AttitudeControl_Multi : public AC_AttitudeControl {
     // calculate total body frame throttle required to produce the given earth frame throttle
     float get_throttle_boosted(float throttle_in);
 
+    // calculate the harmonic notch filter frequency scaled on motor rpm
+    virtual float get_notch_freq_scaled(float ref_freq, float ref) const override;
```

Comment:
it's OK to remove the "virtual" from the child methods.  Having it virtual here means that another level of child classes could also override the function.  It's saves a tiny bit of program space to leave out the unnecessary virtual.

As mentioned in another comment, the get_notch_freq_scaled is such a small function and it doesn't use any of the state or features of AC_AttitudeControl that I wonder if it might be simpler to have it in Copter::update_dynamic_notch() or in the ins library.

Reply:
I'd have to check my history but pretty sure it was moved here from the IMU code. Also we may have to introduce some further shaping on the output related to the thrust expo, which I think naturally fits here - @lthall might have an opinion?

Reply:
Also I think @bnsgeyer wants to do something slightly different for Heli's

Reply:
Sorry to jump on this band wagon late.  I was looking through this and thinking how I would implement this for heli.  We would want to use measured rotor speed.  @andyp1per is the harmonic notch calling the attitude controller to get the scaled frequency?  the reason I ask is that I may not set the scaled frequency in the attitude controller because I'm not using throttle.  Why not have a set function in the harmonic notch that lets anything set the scaled frequency.  that way it is not tied to just the attitude controller.

Reply:
@bnsgeyer currently you have both :) The attitude controller is asked what the frequency should be and then the result is passed into the INS code to request a particular frequency. So you can miss out step 1 and do whatever you like. That was my point to @rmackay9 - it makes sense to keep these things separate rather than rolling them all into one thing because people will want to compose them differently.

Reply:
I just don't see why the get_notch_freq_scaled() function is in the attitude controller.  It doesn't make any use of any state in the attitude controller.  I think the intention is to allow the function to be different based on the vehicle type but there are a number of ways to do it - we don't need to use the attitude controller's vehicle specific class structure to do it.

Reply:
@rmackay9  I had implemented the tradheli to use RPM.  the commit in my repo of this branch is [here](https://github.com/bnsgeyer/ardupilot/commit/f85aaf0b8c9505640991b137ec760e9761816128).  Not sure if you saw this and the discussion Andy and I had on this.

Reply:
@bnsgeyer, yes, maybe I responded in the wrong place.

I'm away for a few days this week at a S&R event so I will need to leave you guys to try and work it out.   ... but I feel pretty strongly that get_notch_freq_scaled() is in the wrong place or perhaps more accurately that the calls it makes to the ins and/or attitude controller are in the wrong place.  I just think it's odd that we are pulling data out of the ins, passing it to a function in the attitude controller that then pulls data from the motors library (but doesn't actually use anything from the attitude controller), and then passes it back to the ins.  This seems like a very roundabout way to do it so surely it can be simplified.

Reply:
@rmackay9 ok, I have moved the calculation to Copter::update_dynamic_notch() and removed the compensation_gain sanity check at @lthall 's request.
@bnsgeyer after talking with @lthall we think it's best if you keep the tradheli change in a separate PR as it needs some further massaging and we don't want to hold up this change going in.

Reply:
Lets see how we go with Bill's heli work over the next week. With your latest changes they may be able to go in pretty easy.